DirectProgramming/DPC++FPGA/include/pipe_utils.hpp

```diff
@@ -0,0 +1,242 @@
+//==============================================================
+// Copyright Intel Corporation
+//
+// SPDX-License-Identifier: MIT
+// =============================================================
+#ifndef __PIPE_ARRAY_HPP__
+#define __PIPE_ARRAY_HPP__
+
+#include <CL/sycl.hpp>
+#include <CL/sycl/INTEL/fpga_extensions.hpp>
+#include <utility>
+
+/*
+
+This header defines the following utilities for use with pipes in DPC++ FPGA
+designs.
+
+1. PipeArray
+
+     Create a collection of pipes that can be indexed like an array.
+
+     template <class Id,          // identifier for the pipe array
+               typename BaseTy,   // type to write/read for each pipe
+               size_t min_depth,  // minimum capacity of each pipe
+               size_t... dims     // depth of each dimension in the array
```

Comment:
Should we drop the ability to have multiple dimensions. @mdbtucker  ?? The code becomes much easier with a single dimension.

Reply:
@tyoungsc why does the code become easier? Isn't the difference from a user's perspective just the number of integers passed as templates (ie. PipeArray<class MyPipe1D, int, 32, 10> vs. PipeArray<class MyPipe2D, int, 32, 5, 2>)?

Reply:
I think we leave it as is.  The internal code gets a lot simpler, but the API for the user doesn't change much.  I don't think we need to care too much about the fact that the internals are complicated.